### PR TITLE
Features/klipper control

### LINKED ISSUE #85: More Buttons and Layerinfos
## Is your feature request related to a problem? Please describe.

No

## Describe your setup

RPi 4 with Moonraker + Mainsail
Anycubic Vyper with Klipper

## Describe the solution you'd like

It would be nice to have Buttons for:
Klipper Control:
 - Restart
 - Firmware Restart

Print Control:
 - Pause
 - Resume
 - Cancel

Host Control:
 - Restart

Layer Infos (If possible):
 - Current Layer
 - Total Layers

### PR DESCRIPTION
**Pull request by cashew22**
*** After #88

Fix #85 

- Add a button for print pause/resume/cancel
- Add a button for restart, host restart and firware restart
- Add total_layer, current_layer and toolhead position

I thought that adding the layer would be easy but the moonraker documentation state that to have those information in the print stats you need to sell your soul to the devil and give your first new born:

```
This is a dict containing information about the print provided by the slicer. Currently this is limited to the total_layer and current_layer values. Note that these values are set by the [SET_PRINT_STATS_INFO](https://www.klipper3d.org/G-Codes.html#set_print_stats_info) gcode command. It is necessary to configure the slicer to include this command in the print. SET_PRINT_STATS_INFO TOTAL_LAYER=total_layer_count should be called in the slicer's "start gcode" to initalize the total layer count. SET_PRINT_STATS_INFO CURRENT_LAYER=current_layer should be called in the slicer's "on layer change" gcode. The user must substitute the total_layer_count and current_layer with the appropriate "placeholder syntax" for the slicer.
```

I decided to with a calculated method that seems to be the same thing as mainsail is doing. To do so, I needed to have the toolhead position in Z and thus this is why I added to three position even so I think those are not that usefull.

### CODE REVIEW

File: custom_components/moonraker/sensor.py [hunk 1]
```diff
@@ -452,3 +498,22 @@ def convert_time(time_s):
     return (
         f"{round(time_s // 3600)}h {round(time_s % 3600 // 60)}m {round(time_s % 60)}s"
     )
+
+
+def calculate_current_layer(data):
```
Comment (marcolivierarsenault) [inline custom_components/moonraker/sensor.py:503]: Nit pick. I would move the function in the class Like `empty_result_when_not_printing`

File: custom_components/moonraker/sensor.py [hunk 2]
```diff
@@ -452,3 +498,22 @@ def convert_time(time_s):
...
+    """Calculate current layer"""
+
+    if (
```
Comment (marcolivierarsenault) [inline custom_components/moonraker/sensor.py:506]: I would call `empty_result_when_not_printing` instead

File: custom_components/moonraker/sensor.py [hunk 3]
```diff
@@ -452,3 +498,22 @@ def convert_time(time_s):
...
+        data["status"]["print_stats"]["state"] != PRINTSTATES.PRINTING.value
+        or data["status"]["print_stats"]["filename"] == ""
+    ):
+        return ""
+
+    # layer = (current_z - first_layer_height) / layer_height + 1
+    return (
+        int(
```
Comment (marcolivierarsenault) [inline custom_components/moonraker/sensor.py:514]: is there a rounding needed? Especially with the bed mesh, I could see the z position to be slightly high or down.

Comment (marcolivierarsenault) [reply custom_components/moonraker/sensor.py:514]: now int probably floor the value

**Review by marcolivierarsenault — COMMENTED** [review]
The way to calculate layer is very smart. Good job.

Did not review the test, waiting for the rebase first

File: custom_components/moonraker/sensor.py (hunk 1 above)
Comment (cashew22) [reply custom_components/moonraker/sensor.py:503]: I don't know, it seems that it fit with the other "calculate" method like eta and pct_job

File: custom_components/moonraker/sensor.py (hunk 3 above)
Comment (cashew22) [reply custom_components/moonraker/sensor.py:514]: The cast to int handle it

**Review by marcolivierarsenault — APPROVED** [review]
Nice

**Pull request merged**